crates/ruff_db/src/file_system/memory.rs

```diff
@@ -1,136 +1,473 @@
+
+        create_dir_all(&mut by_path, &normalized)
     }
 }
 
 impl FileSystem for MemoryFileSystem {
     fn metadata(&self, path: &FileSystemPath) -> Result<Metadata> {
-        let entry = self
-            .inner
-            .files
-            .get(&path.to_path_buf())
-            .ok_or_else(|| std::io::Error::new(ErrorKind::NotFound, "File not found"))?;
-
-        let value = entry.value();
-
-        Ok(Metadata {
-            revision: value.last_modified.into(),
-            permissions: Some(value.permission),
-            file_type: FileType::File,
-        })
+        let by_path = self.inner.by_path.read().unwrap();
+        let normalized = normalize_path(path, &self.inner.cwd);
+
+        let entry = by_path.get(&normalized).ok_or_else(not_found)?;
+
+        let metadata = match entry {
+            Entry::File(file) => Metadata {
+                revision: file.last_modified.into(),
+                permissions: Some(Self::PERMISSION),
+                file_type: FileType::File,
+            },
+            Entry::Directory(directory) => Metadata {
+                revision: directory.last_modified.into(),
+                permissions: Some(Self::PERMISSION),
+                file_type: FileType::Directory,
+            },
+        };
+
+        Ok(metadata)
     }
 
     fn read(&self, path: &FileSystemPath) -> Result<String> {
-        let entry = self
-            .inner
-            .files
-            .get(&path.to_path_buf())
-            .ok_or_else(|| std::io::Error::new(ErrorKind::NotFound, "File not found"))?;
+        let by_path = self.inner.by_path.read().unwrap();
+        let normalized = normalize_path(path, &self.inner.cwd);
 
-        let value = entry.value();
+        let entry = by_path.get(&normalized).ok_or_else(not_found)?;
 
-        Ok(value.content.clone())
+        match entry {
+            Entry::File(file) => Ok(file.content.clone()),
+            Entry::Directory(_) => Err(is_a_directory()),
+        }
     }
 
     fn exists(&self, path: &FileSystemPath) -> bool {
-        self.inner.files.contains_key(&path.to_path_buf())
+        let by_path = self.inner.by_path.read().unwrap();
+        let normalized = normalize_path(path, &self.inner.cwd);
+
+        by_path.contains_key(&normalized)
+    }
+}
+
+impl Default for MemoryFileSystem {
+    fn default() -> Self {
+        MemoryFileSystem::new()
     }
 }
 
 impl std::fmt::Debug for MemoryFileSystem {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        let mut map = f.debug_map();
+        let paths = self.inner.by_path.read().unwrap();
 
-        for entry in self.inner.files.iter() {
-            map.entry(entry.key(), entry.value());
-        }
-        map.finish()
+        f.debug_map().entries(paths.iter()).finish()
     }
 }
 
-#[derive(Default)]
 struct MemoryFileSystemInner {
-    files: FxDashMap<FileSystemPathBuf, FileData>,
+    by_path: RwLock<FxHashMap<Utf8PathBuf, Entry>>,
```

Comment:
Why the switch from `FxDashMap` to an `FxHashMap` wrapped in a `RwLock`?

Reply:
Good question. I don't think it was necessary as part of this PR. I first thought I would need a `BTreeMap`. I'll leave it this way because I do need a `BTreeMap` when adding support for removing directories.